### ISSUE TITLE
chaincode execution should prevent accidental sharing of global state in chaincode

### ISSUE THREAD

**Issue by corecode** [hyperledger-archives/fabric#1751]
## Description

Currently, the chaincode pattern requires passing of a singleton object to the chaincode shim:

```
    err := shim.Start(new(SimpleChaincode))
```

This suggests that it is a reasonable practice to use a singleton object to serve Invokes and Queries.  As soon as these objects carry state, i.e. they are anything but `struct{}`, this leads to incorrectly sharing state between concurrent executions of the chaincode.  This behavior should be prevented, not encouraged.
### Change to factory function

If we change the interface to take a factory function, instead of an object, then a new object can be created for every invocation:

```
shim.Start(newSimpleChaincode)
```

However, this approach still does not prevent accidental sharing of data through other singletons or global variables.
### Isolation of Execution

We can isolate execution more thoroughly by just executing a new chaincode process on every query or invoke.  The OS provides standard isolation guarantees.  There is no accidental sharing of state; only deliberate sharing (via the file system or System V shared memory) is still possible.  This deliberate sharing can be further curbed by installing syscall filters.

This separation into processes can either be handled from the fabric side (executing a new process for every invoke/query), or by the chaincode shim (provide Main by shim, shim forks for every new query/execute and then passes execution to the chaincode).
## How to reproduce
#1545 most likely due to accidental chaincode non-determinism.  There have been several other instances of accidental sharing in chaincode as well (handled via slack).

**Comment by corecode** [comment]
Right, it is #1545. Corrected above.

**Comment by ghaskins** [comment]
Regarding Isolation, I have been thinking along similar lines myself.  I went so far as to build an experiment to see how fast a golang process could be fork/exec'd.  I'll see if I can dig it up and post here.

**Comment by sachikoy** [comment]
I agree isolation is necessary, but if we start a new OS process at every chain code execution, the performance overhead may be too large. Alternatively, is it possible to verify the chain code on its deployment and reject if it uses any global variables?
